scripts/staging/tSNE.dml

```diff
@@ -0,0 +1,131 @@
+X = read($X)
+
+distance_matrix = function(matrix[double] X)
+  return (matrix[double] out) {
+    n = nrow(X)
+    s = rowSums(X * X)
+    out = - 2*X %*% t(X) + s + t(s)
+  }
+
+x2p = function(matrix[double] X, double tol, double perplexity)
+  return(matrix[double] P) {
+    INF = 1.0e20
+    n = nrow(X)
+    D = distance_matrix(X)
+
+    P = matrix(0, rows=n, cols=n)
+    beta = matrix(1, rows=n, cols=1)
+    logU = log(perplexity)
+
+    for (i in 1:n) {
+      betamin = 0 # beta cannot be negative
+      betamax = INF
+      Hdiff = INF
+      itr = 0
+
+      while (abs(Hdiff) > tol & itr < 50) {
+        P[i,] = exp(-D[i,] * beta[i,1])
+        P[i,i] = 0.
+        sum_Pi = sum(P[i,])
+
+        H = log(sum_Pi) + beta[i,1] * sum(D[i,] * P[i,]) / sum_Pi
+        P[i,] = P[i,]/sum_Pi
+
+        Hdiff = as.scalar(H - logU)
+        # print(i + " " + itr + " " +  as.scalar(D[i,1]) + " " + as.scalar(beta[i,1]) +
+        #   " " + as.scalar(P[i,1]) + " " + sum_Pi + " " + as.scalar(H[1,1]))
+        if (Hdiff > 0.) {
+          betamin = as.scalar(beta[i,1])
+          if (betamax == INF) {
+            beta[i,1] = beta[i,1] * 2.
+          } else {
+             beta[i,1] = (beta[i,1] + betamax) / 2.
+          }
+        } else {
+          betamax = as.scalar(beta[i,1])
+          if (betamin == 0.) {
+            beta[i,1] = beta[i,1] / 2.
+          } else {
+            beta[i,1] = (beta[i,1] + betamin) / 2.
+          }
+        }
+        itr = itr + 1
+      }
+    }
+    P = P + t(P)
+    P = P / sum(P)
+  }
+
+tsne = function(matrix[double] X, int reduced_dims, int initial_dims, int perplexity)
+  return(matrix[double] Y, matrix[double] C) {
+    d = reduced_dims
+    n = nrow(X)
+
+    max_iter = 2000
+    eta = 500
+
+    P = x2p(X, 1.0e-5, 20.0)
+    P = P*4
+    Y = rand(rows=n, cols=d, pdf="normal")
+    C = matrix(0, rows=max_iter, cols=1)
+
+    for (itr in 1:max_iter) {
+      D = distance_matrix(Y)
+      Z = 1/(D + 1)
+      parfor (j in 1:n) {
```

Comment:
Thanks Imran. Minor suggestion when you run performance experiments: You can vectorize the parfor which might help improve the performance.
``` python
Z = Z * (diag(matrix(-1, rows=nrow(Z), cols=1)) + 1)
```